### PR TITLE
fix: iterative variable assignments in AutoQASM

### PR DESCRIPTION
**Pull request by rmshaffer**
*Issue #, if available:*
https://github.com/amazon-braket/amazon-braket-sdk-python/issues/929, https://github.com/amazon-braket/amazon-braket-sdk-python/issues/933

*Description of changes:*
Ensure that AutoQASM variables which are updated after declaration, including iteratively in a loop, are initialized properly.

*Testing done:*
Added tests for new cases. Updated existing tests where variable declaration order changed. `tox` passes.

*Note:*
If a variable is initialized with a Python variable and then updated iteratively in a loop, this scenario still does not work with the changes in this PR. A new issue has been added to the backlog for this: https://github.com/orgs/amazon-braket/projects/2/views/1?pane=issue&itemId=58435320.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/experimental/autoqasm/program/program.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
 import oqpy.base
 import pygments
+from openpulse import ast
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/program/program.py:28]: Why openpulse?

File: src/braket/experimental/autoqasm/program/program.py [hunk 2]
```diff
@@ -509,10 +510,23 @@ def add_io_declarations(self) -> None:
...
                 root_oqpy_program._add_var(parameter)
+
         for parameter_name, parameter in self._output_parameters.items():
             # Before adding the output variable to the program, remove any existing reference
-            root_oqpy_program.undeclared_vars.pop(parameter_name, None)
-            root_oqpy_program.declared_vars.pop(parameter_name, None)
+            popped_undeclared = root_oqpy_program.undeclared_vars.pop(parameter_name, None)
+            popped_declared = root_oqpy_program.declared_vars.pop(parameter_name, None)
+
+            popped = popped_undeclared if popped_undeclared is not None else popped_declared
```
Comment (laurencap) [inline src/braket/experimental/autoqasm/program/program.py:519]: Because we know we won't see one parameter in both lists, right?

**Review by laurencap — APPROVED** [review]

File: src/braket/experimental/autoqasm/program/program.py (hunk 1 above)
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/program/program.py:28]: This is a superset of openqasm3 ast - openpulse just imports and re-exports the entire openqasm3 ast.
https://github.com/openqasm/openpulse-python/blob/b3a4ae3faf855527936130245ea2d7620d199f2a/source/openpulse/openpulse/ast.py#L17-L19

In our codebase we just tend to import openpulse ast everywhere. Probably not strictly needed here.

File: src/braket/experimental/autoqasm/program/program.py (hunk 2 above)
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/program/program.py:519]: That should certainly be the case. But it's a good point, I'll add an assert here to verify that we never have both.

File: src/braket/experimental/autoqasm/operators/assignments.py [hunk 3]
```diff
@@ -170,17 +171,17 @@ def assign_stmt(target_name: str, value: Any) -> Any:
         # For example:
         #   int[32] a = 10;
         # where `a` is at the root scope of the function (not inside any if/for/while block).
-        target.init_expression = value.init_expression
-        oqpy_program.declare(target)
+        target.init_expression = value_init_expression
+        oqpy_program._add_var(target)
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/operators/assignments.py:175]: Good catch! Thanks, I'll follow up on this and get it fixed.

[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/braket/experimental/autoqasm/operators/assignments.py:175]: This is being addressed in https://github.com/amazon-braket/amazon-braket-sdk-python/pull/938

**Pull request merged**